BSP/STM32F413/Src/BSP_UART.c

```diff
@@ -44,208 +67,598 @@ void BSP_UART_Init(void) {
-    if(!TxFifo_IsFull()) {
-        txBuffer[txPut] = data;
-        txPut = (txPut + 1) % TX_SIZE;
+static bool RxFifo_RemoveLast(uint8_t *data, UART_Port usart) {
+    if(!RxFifo_IsEmpty(UART_USB) && (usart == UART_USB)) {
+        *data = rxBuffer3[rxPut3 - 1];
+        if(rxPut3 > 0) rxPut3--;
+        else rxPut3 = RX_SIZE - 1;
+        return true;
+    }
+    if(!RxFifo_IsEmpty(UART_BLE) && (usart == UART_BLE)) {
+        *data = rxBuffer2[rxPut2 - 1];
+        if(rxPut2 > 0) rxPut2--;
+        else rxPut2 = RX_SIZE - 1;
         return true;
     }
+    return false;
+}
 
+static bool RxFifo_Peek(uint8_t *data, UART_Port usart) {
+    if(!RxFifo_IsEmpty(UART_USB) && (usart == UART_USB)) {
+        *data = rxBuffer3[rxGet3];
+        return true;
+    }
+    if(!RxFifo_IsEmpty(UART_BLE) && (usart == UART_BLE)) {
+        *data = rxBuffer2[rxGet2];
+        return true;
+    }
     return false;
 }
 
-static bool TxFifo_IsFull(void) {
-    return (txPut + 1) % TX_SIZE == txGet;
+static bool RxFifo_IsFull(UART_Port usart) {
+    if (usart == UART_USB) return (rxPut3 + 1) % RX_SIZE == rxGet3;
+    if (usart == UART_BLE) return (rxPut2 + 1) % RX_SIZE == rxGet2;
+    return 0;
 }
 
-static bool TxFifo_IsEmpty(void) {
-    return txGet == txPut;
+static bool RxFifo_IsEmpty(UART_Port usart) {
+    if (usart == UART_USB) return rxGet3 == rxPut3;
+    if (usart == UART_BLE) return rxGet2 == rxPut2;
+    return 0;
+}
+#endif
+//**********THE FOLLOWING CODE IS FOR THE RTOS VERSION OF THE BPS**********
+#ifdef RTOS
+/**
+ * @brief   Gets one line of ASCII text that was received.
+ * @pre     str should be at least 128bytes long.
+ * @param   str : pointer to buffer to store the string. This buffer should be initialized
+ *                  before hand.
+ * @param   usart : which usart to read from (2 or 3)
+ * @return  number of bytes that was read
+ */
+uint32_t BSP_UART_ReadLine(char *str, UART_Port usart) {
+    uint8_t data = 0;
+    uint32_t recvd = 0;
+    //lineRecieved3 is used by rx and tx but only rx mutex will be used for access
+    OSMutexPend(&rx3Mutex, 0, OS_OPT_PEND_BLOCKING, &time, &err);
+    if(lineReceived3 && (usart == UART_USB)) { //read from 3rd usart
+        OSMutexPost(&rx3Mutex, OS_OPT_POST_NONE, &err);
+        USART_ITConfig(USART3, USART_IT_RXNE, RESET);
+        RxFifo_Peek(&data, UART_USB);
+        while(!RxFifo_IsEmpty(UART_USB) && data != '\r') {
+            recvd += RxFifo_Get((uint8_t *)str++, UART_USB);
+            RxFifo_Peek(&data, UART_USB);
+        }
+        RxFifo_Get(&data, UART_USB);
+        *str = 0;
+        OSMutexPend(&rx3Mutex, 0, OS_OPT_PEND_BLOCKING, &time, &err);
+        lineReceived3 = false;
+        OSMutexPost(&rx3Mutex, OS_OPT_POST_NONE, &err);
+        USART_ITConfig(USART3, USART_IT_RXNE, SET);
+        return recvd;
+    }
+    //lineRecieved2 is used by rx and tx but only rx mutex will be used for access
+    OSMutexPend(&rx2Mutex, 0, OS_OPT_PEND_BLOCKING, &time, &err);
```

Comment:
What happens if the statement `if(lineReceived2 && (usart == UART_BLE))` is false and we return without posting the mutex? I think this could cause UART to freeze.